src/goto-checker/all_properties_verifier_with_fault_localization.h

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************\
+
+Module: Goto Verifier for Verifying all Properties and Localizing Faults
+
+Author: Daniel Kroening, Peter Schrammel
+
+\*******************************************************************/
+
+/// \file
+/// Goto verifier for verifying all properties that stores traces
+/// and localizes faults
+
+#ifndef CPROVER_GOTO_CHECKER_ALL_PROPERTIES_VERIFIER_WITH_FAULT_LOCALIZATION_H
+#define CPROVER_GOTO_CHECKER_ALL_PROPERTIES_VERIFIER_WITH_FAULT_LOCALIZATION_H
+
+#include "goto_verifier.h"
+
+#include "fault_localization_provider.h"
+#include "goto_trace_storage.h"
+#include "incremental_goto_checker.h"
+#include "properties.h"
+#include "report_util.h"
+
+/// Requires an incremental goto checker that is a
+/// `goto_trace_providert` and `fault_localization_providert`.
+template <class incremental_goto_checkerT>
+class all_properties_verifier_with_fault_localizationt : public goto_verifiert
+{
+public:
+  all_properties_verifier_with_fault_localizationt(
+    const optionst &options,
+    ui_message_handlert &ui_message_handler,
+    abstract_goto_modelt &goto_model)
+    : goto_verifiert(options, ui_message_handler),
+      goto_model(goto_model),
+      incremental_goto_checker(options, ui_message_handler, goto_model),
+      traces(incremental_goto_checker.get_namespace())
+  {
+    properties = initialize_properties(goto_model);
+  }
+
+  resultt operator()() override
+  {
+    while(true)
+    {
+      const auto result = incremental_goto_checker(properties);
+      if(result.progress == incremental_goto_checkert::resultt::progresst::DONE)
+        break;
+
+      // we've got an error trace
+      message_building_error_trace(log);
+      for(const auto &property_id : result.updated_properties)
+      {
+        if(properties.at(property_id).status == property_statust::FAIL)
+        {
+          // get correctly truncated error trace for property and store it
+          (void)traces.insert(
```

Comment:
Are these both expected to be insert-new-key? If so suggest cross-checking with a CHECK_RETURN that they are indeed new.

Reply:
(If not, suggest inserting a dummy value and only calling the expensive build / localize functions if they really are new)

Reply:
I've added an invariant in the trace storage.